File: sync.py
"""
Organizer — the GitHub buttons.

The Mac and the PC pass this folder between them through a GitHub
repository. This file is the only part of Organizer that talks to it, and
it does exactly two things, each only when asked:

    Get    bring in what the other computer sent
    Send   commit what changed here, and push it

Everything leans on the rule the README already keeps: one computer at a
time. Get before you start, Send before you stop.

WHAT IT WILL NOT DO

    Get only ever fast-forwards. If both computers changed the same file,
    git would stitch the two versions of data.json together line by line
    and could leave something that isn't valid JSON at all. So instead of
    guessing, this refuses and names the files that are in the way.

    Nothing waits on a password prompt. Organizer usually runs with no
    Terminal attached, so a prompt would hang the server with nobody there
    to answer it. git is told never to ask, and a missing login comes back
    as a message saying how to sign in once.

    Files too big for GitHub — it refuses anything of 100 MB or more — are
    left out of a Send and named. Committing one would put it into a
    history that could then never be pushed, and every Send after it
    would fail for a reason that had nothing to do with that Send.
"""

import glob
import json
import os
import re
import shutil
import subprocess
import threading
import time

ROOT = os.path.dirname(os.path.abspath(__file__))

# GitHub rejects any file of 100 MB or more. Stay a little clear of it.
TOO_BIG = 95 * 1024 * 1024

# Long enough for a slow connection, or a first sign-in window on Windows.
NETWORK_SECONDS = 90
LOCAL_SECONDS = 20

# When a Get brings in changes to these, the page can reload but the
# server running it is still the old one — it has to be restarted.
SERVER_FILES = ("server.py", "launch.py", "sync.py")
SERVER_DIRS = ("desktop/",)

# One git operation at a time. git guards itself with a lock file and
# fails outright if two run at once — say, the check that runs when the
# window comes back into focus, landing on top of a Send.
_lock = threading.Lock()


# ============================================================
#  RUNNING GIT
# ============================================================
def find_git():
    """The git to use, or None."""
    found = shutil.which("git")
    if found:
        return found
    # GitHub Desktop carries its own git, so a machine with only Desktop
    # installed still has one — just not on the PATH.
    places = [
        "/Applications/GitHub Desktop.app/Contents/Resources/app/git/bin/git",
        os.path.expanduser("~/Applications/GitHub Desktop.app/Contents/Resources/app/git/bin/git"),
    ]
    local = os.environ.get("LOCALAPPDATA")
    if local:
        pattern = os.path.join(local, "GitHubDesktop", "app-*", "resources",
                               "app", "git", "cmd", "git.exe")
        places += sorted(glob.glob(pattern), key=os.path.getmtime, reverse=True)
    return next((p for p in places if os.path.exists(p)), None)


def _git(*args, timeout=LOCAL_SECONDS):
    """(exit code, stdout, stderr). Never prompts, never hangs past `timeout`."""
    git = find_git()
    if not git:
        return 127, "", "git not found"

    env = dict(os.environ)
    env["GIT_TERMINAL_PROMPT"] = "0"            # never stop and ask for a password
    env.setdefault("GIT_SSH_COMMAND", "ssh -o BatchMode=yes")
    env.pop("GIT_ASKPASS", None)
    env.pop("SSH_ASKPASS", None)

    extra = {}
    if os.name == "nt":
        # Organizer runs windowless on the PC; without this every git call
        # would flash a black console window across the screen.
        extra["creationflags"] = subprocess.CREATE_NO_WINDOW

    try:
        done = subprocess.run(
            [git, "-C", ROOT, "-c", "core.quotepath=off", *args],
            capture_output=True, text=True, encoding="utf-8", errors="replace",
            env=env, timeout=timeout, **extra)
        return done.returncode, done.stdout, done.stderr
    except subprocess.TimeoutExpired:
        return 124, "", "timed out"
    except OSError as err:
        return 127, "", str(err)


def _last_line(text: str) -> str:
    lines = [l.strip() for l in (text or "").splitlines() if l.strip()]
    line = lines[-1] if lines else ""
    # A remote URL can carry a login inside it. Never hand one to the page.
    return re.sub(r"://[^/@\s]+@", "://", line)[:200]


def _why(err: str) -> str:
    """Sort a failed network call into something the page can explain."""
    e = (err or "").lower()
    if any(s in e for s in ("could not read username", "terminal prompts disabled",
                            "authentication failed", "invalid username or password",
                            "permission denied", "the requested url returned error: 403",
                            "repository not found")):
        # A private repository answers "not found" to anyone not signed in,
        # rather than admitting it exists. So this is almost always a login.
        return "login"
    if any(s in e for s in ("could not resolve host", "failed to connect",
                            "network is unreachable", "timed out", "no route to host",
                            "connection refused", "connection reset")):
        return "offline"
    return "other"


# ============================================================
#  WHAT'S CHANGED
# ============================================================
def _entries():
    """[(status, path)] for every change git can see, ignored files aside.

    -z so a path comes back exactly as it is on disk. Without it git quotes
    anything unusual, and a project called  Videoclip — “Nightdrive”  turns
    into a string of octal escapes that matches no file at all.
    """
    code, out, _ = _git("status", "--porcelain=v1", "-z", "--untracked-files=all")
    if code != 0:
        return []
    parts, found, i = out.split("\0"), [], 0
    while i < len(parts):
        item = parts[i]
        i += 1
        if len(item) < 4:
            continue
        xy, path = item[:2], item[3:]
        if xy[0] in "RC":          # a rename also lists where it came from
            i += 1
        found.append((xy, path))
    return found


def _data_ok() -> bool:
    """Is data.json something the app can read? (No file at all is fine.)

    Checked before every Send. A merge that stopped half-way leaves conflict
    markers in it, and to git that's just another changed file, ready to
    commit. Sending it would hand the damage to the other computer too.
    """
    path = os.path.join(ROOT, "data.json")
    if not os.path.exists(path):
        return True
    try:
        with open(path, "r", encoding="utf-8") as fh:
            json.load(fh)
        return True
    except (OSError, ValueError):
        return False


def _size(path: str) -> int:
    try:
        return os.path.getsize(os.path.join(ROOT, path))
    except OSError:
        return 0                   # deleted: sending a deletion costs nothing


# ============================================================
#  STATUS
# ============================================================
def status(fetch=False):
    with _lock:
        return _status(fetch)


def _status(fetch: bool) -> dict:
    out = {"git": False}

    if not find_git():
        out["reason"] = "no-git"
        return out

    code, top, _ = _git("rev-parse", "--show-toplevel")
    same = code == 0 and (os.path.normcase(os.path.realpath(top.strip()))
                          == os.path.normcase(os.path.realpath(ROOT)))
    if not same:
        out["reason"] = "not-a-repo"
        return out

    code, upstream, _ = _git("rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}")
    if code != 0:
        out["reason"] = "no-upstream"
        return out
    upstream = upstream.strip()                    # e.g. origin/main
    out.update(git=True, upstream=upstream)

    if fetch:
        code, _, err = _git("fetch", "--quiet", upstream.split("/", 1)[0],
                            timeout=NETWORK_SECONDS)
        if code == 0:
            out["checkedAt"] = int(time.time() * 1000)
        else:
            out["error"] = _why(err)
            out["detail"] = _last_line(err)

    code, counts, _ = _git("rev-list", "--left-right", "--count", "HEAD...@{u}")
    ahead, behind = (int(n) for n in counts.split()) if code == 0 else (0, 0)
    out["ahead"], out["behind"] = ahead, behind

    entries = _entries()
    changed = [p for _, p in entries]
    out["tooBig"] = [p for p in changed if _size(p) >= TOO_BIG]
    out["changed"] = [p for p in changed if p not in out["tooBig"]]

    # What the other computer sent since the two last agreed, and whether
    # any of it lands on a file that's also been changed here — the one
    # thing a fast-forward can't get past.
    incoming = []
    if behind:
        code, names, _ = _git("diff", "--name-only", "-z", "HEAD...@{u}")
        incoming = [n for n in names.split("\0") if n]
    edited_here = {p for xy, p in entries if xy != "??"}
    out["incoming"] = incoming
    out["clash"] = sorted(edited_here & set(incoming))

    out["broken"] = not _data_ok()
    out["canGet"] = bool(behind) and not ahead and not out["clash"]
    out["canSend"] = bool(out["changed"] or ahead) and not behind and not out["broken"]

    code, head, _ = _git("log", "-1", "--format=%s%x00%ct", "@{u}")
    if code == 0 and "\0" in head:
        subject, when = head.strip().split("\0", 1)
        out["remote"] = {"subject": subject, "when": int(when or 0) * 1000}
    return out


# ============================================================
#  GET
# ============================================================
def get() -> dict:
    with _lock:
        st = _status(fetch=True)
        if not st.get("git") or st.get("error"):
            return {**st, "ok": False}
        if not st["behind"]:
            return {**st, "ok": True, "got": []}
        if st["ahead"]:
            return {**st, "ok": False, "why": "diverged"}
        if st["clash"]:
            return {**st, "ok": False, "why": "clash"}

        _, before, _ = _git("rev-parse", "HEAD")
        # --ff-only: move forward to what GitHub has, or do nothing at all.
        # Never a merge commit, never a file overwritten that was changed here.
        code, _, err = _git("merge", "--ff-only", "--quiet", "@{u}")
        if code != 0:
            return {**st, "ok": False, "why": "refused", "detail": _last_line(err)}

        _, names, _ = _git("diff", "--name-only", "-z", before.strip(), "HEAD")
        got = [n for n in names.split("\0") if n]
        restart = any(n in SERVER_FILES or n.startswith(SERVER_DIRS) for n in got)
        return {**_status(fetch=False), "ok": True, "got": got, "restart": restart}


# ============================================================
#  SEND
# ============================================================
def _message(device: str, files: list) -> str:
    stamp = time.strftime("%a %d %b %Y, %H:%M")
    head = f"Sent from {device} · {stamp}"
    shown = files[:25]
    body = "\n".join(f"  {f}" for f in shown)
    more = f"\n  …and {len(files) - len(shown)} more" if len(files) > len(shown) else ""
    return f"{head}\n\n{len(files)} file{'s' if len(files) != 1 else ''}:\n{body}{more}"


def _identity(device: str) -> list:
    """Commit as whoever git is set up as, or as this machine if nobody is.

    A PC with GitHub Desktop but no git of its own may have no name set at
    all, and git refuses to commit without one.
    """
    extra = []
    if not _git("config", "user.name")[1].strip():
        extra += ["-c", f"user.name=Organizer on {device}"]
    if not _git("config", "user.email")[1].strip():
        extra += ["-c", "user.email=organizer@localhost"]
    return extra


def send(device: str) -> dict:
    with _lock:
        st = _status(fetch=True)
        if not st.get("git") or st.get("error"):
            return {**st, "ok": False}
        if st["broken"]:
            return {**st, "ok": False, "why": "broken"}
        if st["behind"]:
            # Sending now would mean merging on top of what's arrived. Get
            # first; if the two collide, that's where it'll say so.
            return {**st, "ok": False, "why": "diverged" if (st["ahead"] or st["clash"]) else "behind"}
        if not st["changed"] and not st["ahead"]:
            return {**st, "ok": True, "sent": [], "nothing": True}

        files = list(st["changed"])
        if files:
            code, _, err = _git("add", "-A")
            if code == 0 and st["tooBig"]:
                # Unstage them again: they stay here, untouched, just unsent.
                code, _, err = _git("reset", "-q", "--",
                                    *[":(literal)" + p for p in st["tooBig"]])
            if code == 0:
                code, _, err = _git(*_identity(device), "commit", "-q",
                                    "-m", _message(device, files))
            if code != 0:
                return {**_status(fetch=False), "ok": False, "why": "commit",
                        "detail": _last_line(err)}

        # Pushing: if this fails, the commit stays here and the next Send
        # pushes it. Nothing is lost by trying.
        code, _, err = _git("push", "--quiet", timeout=NETWORK_SECONDS)
        after = _status(fetch=False)
        if code != 0:
            kind = _why(err)
            if "rejected" in err or "non-fast-forward" in err or "fetch first" in err:
                kind = "behind"            # the other computer sent something just now
            return {**after, "ok": False, "error": kind, "detail": _last_line(err),
                    "committed": bool(files)}

        _, short, _ = _git("rev-parse", "--short", "HEAD")
        return {**after, "ok": True, "sent": files, "commit": short.strip()}

File: js/ui/sync.js
/* ============================================================
   ui/sync.js
   The GitHub button: Get what the other computer sent, Send
   what this one changed.

   The work happens in sync.py; this is the part you see. One
   button in the top bar with a small panel under it, and a badge
   counting what this computer hasn't sent yet — the thing to
   glance at before you quit.

   GET happens by itself, when it's safe. On opening, and when
   the window comes back into focus, this asks GitHub whether the
   other computer sent anything. If it did, and nothing here is in
   the way, it's brought in and the page reloads onto it. That's
   the one direction worth automating: working on top of old data
   is how both computers end up changing the same thing, and that
   is the mess nothing can tidy up by itself.

   SEND never happens by itself. You decide when — before quitting,
   usually. Sending every few minutes would bury the history in
   hundreds of commits, most of them work half-done.
   ============================================================ */
window.ORG = window.ORG || {};

ORG.sync = (() => {
  const U = ORG.util;

  const RECHECK = 60 * 1000;        // on focus, ask GitHub at most once a minute
  const LOOP_GUARD = 30 * 1000;     // and never auto-Get twice in a row this fast

  let st = null;                    // the last answer from /api/sync
  let busy = null;                  // "check" | "get" | "send" while one is running
  let note = null;                  // a line about what just happened
  let lastCheck = 0;

  const onWindows = /win/i.test(navigator.platform || navigator.userAgent);

  /* ============================================================
     TALKING TO THE SERVER
     ============================================================ */
  async function call(path, method = "GET", body){
    let res;
    try {
      res = await fetch(path, body === undefined ? { method } : {
        method, headers:{ "Content-Type":"application/json" }, body:JSON.stringify(body),
      });
    }
    catch { return { git:false, reason:"server" }; }
    /* 404 means the server running now predates this file: the page was
       reloaded onto new code, but the server wasn't restarted. */
    if (res.status === 404) return { git:false, reason:"old-server" };
    try { return await res.json(); }
    catch { return { git:false, reason:"server" }; }
  }

  /** Let any save that's pending land on disk first. Otherwise the last
      thing typed is missing from a Send, or — worse — a Get sees this
      computer as clean when it isn't. */
  async function settle(){
    for (let i = 0; i < 40 && ORG.store.busy; i++) await new Promise(r => setTimeout(r, 50));
  }

  /* ============================================================
     WHAT THE STATE MEANS
     ============================================================ */
  const count = (n, one, many = one + "s") => `${n} ${n === 1 ? one : many}`;
  const toSend = s => s.changed.length || s.ahead;
  const diverged = s => s.behind > 0 && (s.ahead > 0 || s.clash.length > 0);

  /* ============================================================
     LINKS TO GITHUB
     ============================================================ */

  /** GitHub's page for a new token, filled in as far as a link can fill it.
      Only the repository itself can't be chosen from a link — so the steps
      say which box to tick. Contents: read and write is all it gets, for
      one year, and it's named after this computer so the two are easy to
      tell apart if one ever needs cancelling. */
  function tokenUrl(){
    const [owner, name] = (st.repo || "/").split("/");
    const device = st.device || "this computer";
    const q = new URLSearchParams({
      name: `Organizer on ${device}`.slice(0, 40),
      description: `Lets Organizer on ${device} get and send ${name}. Nothing else.`,
      target_name: owner,
      expires_in: "366",
      contents: "write",
    });
    return `https://github.com/settings/personal-access-tokens/new?${q}`;
  }

  /** Open a github.com page in the browser you normally use, where you're
      signed in to GitHub — not in this window, which is a Firefox profile
      of its own, signed in to nothing. The server does the opening; if it
      can't (it predates this), fall back to a new window here. */
  async function openOnGitHub(url){
    const r = await call("/api/open", "POST", { url });
    if (!r.ok) window.open(url, "_blank", "noopener");
  }

  /** { tone, word, line, kind } for the current state. */
  function reading(){
    if (busy === "get")  return { tone:"busy", word:"Getting…", line:"Bringing in what the other computer sent." };
    if (busy === "send") return { tone:"busy", word:"Sending…", line:"Committing and pushing what changed here." };
    if (!st)             return { tone:"busy", word:"Checking…", line:"Asking GitHub what's new." };

    if (!st.git){
      switch (st.reason){
        case "old-server": return { tone:"warn", word:"Restart needed",
          line:"Quit and reopen Organizer to switch this on — the server running now is older than the page." };
        case "no-git": return { tone:"off", word:"No git",
          line:"There's no git on this computer. Installing GitHub Desktop brings one." };
        case "not-a-repo": return { tone:"off", word:"Not linked",
          line:"This folder isn't a GitHub repository." };
        case "no-upstream": return { tone:"off", word:"Not linked",
          line:"This folder isn't connected to a repository on GitHub yet. Publish it once from GitHub Desktop." };
        default: return { tone:"off", word:"Unavailable", line:"The server isn't answering." };
      }
    }

    if (st.broken || ORG.store.broken) return { tone:"danger", word:"Can't send",
      line:"data.json is damaged, so nothing will be sent — sending it would pass the damage "
        + "to the other computer. Put it right in GitHub Desktop first, then reload." };

    /* Before anything else that could lead to a Send: while strangers can
       read the repository, nothing goes to it. The server refuses too —
       this is only the explanation. */
    if (st.public) return { tone:"danger", word:"Public", kind:"public",
      line:`Anyone can read ${st.repo || "this repository"} — your tasks, notes and attachments. `
        + "Nothing will be sent until it's private." };

    if (st.error === "login") return { tone:"warn", word:"Sign-in needed", kind:"login",
      line:"Git on this computer isn't signed in to GitHub yet. Once, and it's done:" };
    if (st.error === "offline") return { tone:"off",  word:"Offline",
      line:"Couldn't reach GitHub. Nothing is lost — try again once you're online." };
    if (st.error === "behind")  return { tone:"warn", word:"Get first",
      line:"The other computer sent something just now. Get it, then Send." };
    if (st.error)               return { tone:"warn", word:"GitHub refused",
      line:st.detail || "GitHub didn't accept that." };

    if (diverged(st)) return { tone:"danger", word:"Both changed",
      line:"The other computer sent changes, and this one changed "
        + (st.clash.length ? `the same ${count(st.clash.length, "file")}` : "things too")
        + ". That needs a person to choose — GitHub Desktop can pull and ask which version of each file to keep." };

    if (st.behind)   return { tone:"get", word:`${st.behind} to get`,
      line:"The other computer sent changes." };
    if (toSend(st))  return { tone:"send", word:`${count(toSend(st), "change")} to send`,
      line:"Changed here since the last Send. Send before you quit, so the other computer gets it." };
    return { tone:"ok", word:"Up to date", line:"Both computers have the same work." };
  }

  /** Plain words for a list of paths: tasks, attachments, the app. */
  function describe(files){
    const out = [];
    if (files.includes("data.json")) out.push("Your tasks and notes");
    const att = files.filter(f => f.startsWith("FILES/"));
    if (att.length){
      const names = att.slice(0, 3).map(f => f.split("/").slice(-2).join(" / "));
      out.push(`${count(att.length, "attachment")}: ${names.join(", ")}${att.length > 3 ? "…" : ""}`);
    }
    const app = files.filter(f => f !== "data.json" && !f.startsWith("FILES/"));
    if (app.length) out.push(`${count(app.length, "app file")} — Organizer itself`);
    return out;
  }

  /* ============================================================
     DRAWING IT
     ============================================================ */
  function paint(){
    const r = reading();
    const btn = U.$("#syncbtn");
    btn.dataset.tone = r.tone;
    btn.title = `GitHub — ${r.word}`;
    /* The count shows only when sending is the next thing to do. Anything
       else — something to get, or trouble — is the dot's job instead, so
       the two never sit on top of each other. */
    U.$("#sync-n").textContent = r.tone === "send" ? String(toSend(st)) : "";

    const panel = U.$("#syncpanel");
    if (panel.hidden) return;
    panel.innerHTML = "";

    const head = U.el("div", "lp-head", "GitHub");
    const word = U.el("span", "sp-word", r.word);
    word.dataset.tone = r.tone;
    head.append(word);
    panel.append(head);

    panel.append(U.el("p", "sp-line", note || r.line));

    if (r.kind === "public") panel.append(goPrivate());
    if (r.kind === "login")  panel.append(signIn());

    // Signing in comes first; the buttons would only fail the same way.
    if (r.kind !== "login"){
      const acts = U.el("div", "sp-acts");
      const get = U.el("button", "btn sm", "↓ Get");
      get.title = "Bring in what the other computer sent";
      get.disabled = !!busy || !st || !st.git;
      get.addEventListener("click", () => doGet(false));

      const send = U.el("button", "btn sm primary", "↑ Send");
      send.title = "Commit what changed here and push it to GitHub";
      send.disabled = !!busy || !st || !st.git || !st.canSend || r.kind === "public";
      send.addEventListener("click", doSend);
      acts.append(get, send);
      panel.append(acts);
    }

    // What's waiting, in plain words. Not while data.json is damaged: a
    // list of things "to send" under "Can't send" reads like a promise.
    if (st && st.git && !st.broken && !ORG.store.broken){
      const list = st.behind && st.incoming.length ? describe(st.incoming)
                 : toSend(st) ? describe(st.changed) : [];
      if (list.length){
        const ul = U.el("ul", "sp-files");
        for (const line of list) ul.append(U.el("li", null, line));
        panel.append(ul);
      }
      if (st.tooBig && st.tooBig.length){
        panel.append(U.el("p", "sp-warn",
          `Too big for GitHub, kept on this computer only: ${st.tooBig.map(f => f.split("/").pop()).join(", ")}`));
      }
      if (st.remote && st.remote.subject){
        panel.append(U.el("div", "sp-foot",
          `On GitHub: ${st.remote.subject.split(" · ")[0]} · ${U.fmtWhen(st.remote.when)}`));
      }
    }
  }

  /** Step one of two: make the repository private. */
  function goPrivate(){
    const box = U.el("div", "sp-steps");
    const open = U.el("button", "btn sm", "Open its settings on GitHub");
    open.addEventListener("click", () => openOnGitHub(`https://github.com/${st.repo}/settings`));
    box.append(open);
    box.append(U.el("p", "sp-hint",
      "Right at the bottom, under Danger Zone: Change visibility → Make private. "
      + "Come back here afterwards and it's noticed within a few seconds."));
    return box;
  }

  /* Step two: sign git in. The field survives a redraw — coming back from
     the browser redraws this panel, and that must not throw away a token
     just pasted. The draft lives here only until it's sent, then goes. */
  let draft = "";
  let signing = false;
  let signTrouble = null;

  function signIn(){
    const box = U.el("div", "sp-steps");
    const name = (st.repo || "").split("/")[1] || "this repository";

    const one = U.el("div", "sp-step");
    const oneBody = U.el("div", "sp-body");
    const open = U.el("button", "btn sm", "Open GitHub's token page");
    open.addEventListener("click", () => openOnGitHub(tokenUrl()));
    oneBody.append(open, U.el("p", "sp-hint",
      `It opens in your usual browser. Under Repository access, choose Only select `
      + `repositories and pick ${name} — everything else is already filled in. `
      + `Generate the token and copy it.`));
    one.append(U.el("span", "sp-num", "1"), oneBody);

    const two = U.el("div", "sp-step");
    const twoBody = U.el("div", "sp-body");
    const row = U.el("div", "sp-paste");
    const input = U.el("input", "sp-token");
    input.type = "password";
    input.placeholder = "Paste the token here";
    input.autocomplete = "off";
    input.spellcheck = false;
    input.value = draft;
    input.disabled = signing;
    input.addEventListener("input", () => { draft = input.value; });

    const go = U.el("button", "btn sm primary", signing ? "Checking…" : "Sign in");
    go.disabled = signing;
    const submit = async () => {
      const token = draft.trim();
      if (!token || signing) return;
      signing = true;
      signTrouble = null;
      paint();
      const r = await call("/api/sync/login", "POST", { token });
      draft = "";                          // gone from the page either way
      signing = false;
      if (r.ok){
        await check();
        note = `Signed in as ${r.user}. Send works from now on.`;
      } else {
        signTrouble = trouble(r, name);
      }
      paint();
    };
    go.addEventListener("click", submit);
    input.addEventListener("keydown", e => { if (e.key === "Enter") submit(); });
    row.append(input, go);
    twoBody.append(row);
    if (signTrouble) twoBody.append(U.el("p", "sp-warn", signTrouble));
    two.append(U.el("span", "sp-num", "2"), twoBody);

    box.append(one, two);
    if (!signing) setTimeout(() => { if (draft) input.focus(); }, 0);
    return box;
  }

  function trouble(r, name){
    if (r.reason === "old-server")
      return "Quit and reopen Organizer first — the server running now is older than this form.";
    switch (r.why){
      case "shape":    return "That doesn't look like a GitHub token — it should start with github_pat_. Copy it again.";
      case "rejected": return "GitHub doesn't recognise that token. It's only shown once; if it's lost, make another.";
      case "no-write": return `GitHub knows the token, but it can't send to ${name}. Check that you picked `
                            + `Only select repositories → ${name}, and that Contents says Read and write.`;
      case "offline":  return "Couldn't reach GitHub. Try again once you're online.";
      case "no-store": return onWindows
        ? "Git on this PC has nowhere to keep a login. Installing Git for Windows (git-scm.com) adds one; then sign in again."
        : "The Keychain wouldn't take it. Try once more.";
      case "not-github": return "This folder's repository isn't on GitHub.";
      default: return r.detail || "That didn't work. Try again.";
    }
  }

  /* ============================================================
     DOING IT
     ============================================================ */
  async function check(fresh){
    if (busy) return st;
    busy = "check";
    paint();
    st = await call(`/api/sync?fetch=1${fresh ? "&fresh=1" : ""}`);
    lastCheck = Date.now();
    busy = null;
    note = null;
    paint();
    return st;
  }

  /* After an edit, update the count without asking GitHub anything — the
     badge is what you glance at before quitting, so it can't still be
     saying "up to date" about work done a minute ago. Local only: git
     status on this folder, no network, cheap enough to run after every
     burst of changes. */
  const recount = U.debounce(async () => {
    if (busy || !st || !st.git || ORG.store.broken) return;
    const local = await call("/api/sync");
    if (busy || !local.git) return;
    /* Keep what the last real check learned about the network — a login
       problem, being offline — since this one never went out to ask. */
    st = { ...local, error: st.error, detail: st.detail };
    paint();
  }, 1500);

  function reloadSoon(msg){
    U.toast(msg, 2500);
    setTimeout(() => location.reload(), 700);
  }

  async function doGet(auto){
    if (busy) return;
    await settle();
    busy = "get";
    paint();
    const r = await call("/api/sync/get", "POST");
    busy = null;
    st = r;
    note = null;

    if (r.ok && r.got && r.got.length){
      if (r.restart){
        /* New server code came in. Reloading the page onto it while the
           old server is still running would pair new pages with old
           answers, so stop here and ask for a restart instead. */
        note = "Updated — quit and reopen Organizer to finish. Part of what arrived is the server itself.";
        U.toast("Updated from GitHub — restart Organizer to finish", 6000);
        paint();
        return;
      }
      try { sessionStorage.setItem("org.autoget", String(Date.now())); } catch {}
      reloadSoon(`Brought in ${count(r.got.length, "change")} from GitHub`);
      return;
    }
    if (r.ok && !auto) note = "Already up to date.";
    if (!r.ok && r.why === "clash")
      note = `Not brought in: this computer changed ${r.clash.join(", ")} too, and so did the other one.`;
    paint();
  }

  async function doSend(){
    if (busy) return;
    await settle();
    busy = "send";
    paint();
    const r = await call("/api/sync/send", "POST");
    busy = null;
    st = r;
    note = null;

    if (r.ok && r.nothing) note = "Nothing to send.";
    else if (r.ok){
      note = `Sent ${count(r.sent.length, "file")} to GitHub.`;
      U.toast("Sent to GitHub", 2200);
    }
    else if (r.why === "behind") note = "The other computer sent changes first. Get them, then Send.";
    else if (r.committed) note = "Saved as a commit here, but not sent yet — press Send again to retry.";
    paint();
  }

  /** Ask GitHub, and bring in whatever's safe to. */
  async function refresh(fresh){
    await check(fresh);
    if (!st || !st.git || st.error || !st.canGet) return;

    /* A reload that lands straight back here and Gets again would loop
       forever in a window with no address bar to escape through. */
    let last = 0;
    try { last = +sessionStorage.getItem("org.autoget") || 0; } catch {}
    if (Date.now() - last < LOOP_GUARD) return;
    if (ORG.store.busy) return;          // mid-edit: leave it for the next focus
    doGet(true);
  }

  /* ============================================================
     WIRING
     ============================================================ */
  function toggle(force){
    const panel = U.$("#syncpanel");
    const open = force !== undefined ? force : panel.hidden;
    panel.hidden = !open;
    U.$("#syncbtn").classList.toggle("on", open);
    if (open) paint();
  }

  const cameFrom = (e, ...ids) =>
    (e.composedPath ? e.composedPath() : []).some(n => n && n.id && ids.includes(n.id));

  function showBroken(){
    const b = U.$("#brokenbanner");
    if (b) b.hidden = !ORG.store.broken;
  }

  function init(){
    U.$("#syncbtn").addEventListener("click", e => { e.stopPropagation(); toggle(); });
    document.addEventListener("click", e => {
      if (!U.$("#syncpanel").hidden && !cameFrom(e, "syncpanel", "syncbtn")) toggle(false);
    });
    document.addEventListener("keydown", e => {
      if (e.key === "Escape" && !U.$("#syncpanel").hidden) toggle(false);
    });

    U.$("#broken-reload").addEventListener("click", () => location.reload());
    U.bus.on("broken", () => { showBroken(); paint(); });
    U.bus.on("change", recount);
    showBroken();

    /* Coming back to the window: ask again. Sooner while the repository is
       public — the likeliest reason for being away is making it private,
       and the panel should move on as soon as that's done. */
    const again = () => {
      if (document.visibilityState !== "visible") return;
      const watching = !!(st && st.public);
      if (Date.now() - lastCheck > (watching ? 10 * 1000 : RECHECK)) refresh(watching);
    };
    window.addEventListener("focus", again);
    document.addEventListener("visibilitychange", again);

    paint();
    /* Only when the app is really saving to this folder. In the browser-only
       fallback there's no server to run git, and asking would just fail. */
    if (!ORG.files.onDisk){ st = { git:false, reason:"server" }; paint(); return; }
    /* With data.json damaged, still say where things stand — that's when
       you most need to know — but bring nothing in on top of it. */
    if (ORG.store.broken) check();
    else refresh();
  }

  return { init, check, refresh, get state(){ return st; } };
})();
